### PR TITLE
Adding a chocolatey installer

### PR DESCRIPTION
**Pull request by johnmccrae**
## Description
This PR replaces https://github.com/chef/chef/pull/13603. That PR branch was borked.

This PR comes from a customer request - https://chefio.atlassian.net/browse/INFC-354
We want to provide a way, consistent with other resources, that allows for the admin to install Choco if it is not present.

The main problem here is that we have a number of resources for chocolatey users but customers attempt to use them without having choco installed already and they get weird errors. We did not correctly set customer expectations with the Choco resources that we did with other package managers. The others will typically offer and option to install that package manager if it's missing. Until now, we have not done that with Choco. This improvement corrects that by adding an installer resource for Chocolatey and improves error output to direct the customer to use the installer. We do NOT install Chocolatey for the user as that would deviate from Chef design philosophies around being absolutely transparent about or explicit about what we're doing - meaning we don't perform some action that the customer has not explicitly chosen to do : We don't install the package manager if it's missing, we inform the customer and offer the new resource to fix that issue.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:

- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have run the pre-merge tests locally and they pass.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] If `Gemfile.lock` has changed, I have used `--conservative` to do it and included the full output in the Description above.
- [ ] All new and existing tests passed.
- [ ] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

**Review by tpowell-progress — APPROVED** [review]
Verify in chef-oss and all GH Actions green.

File: lib/chef/resource/chocolatey_installer.rb [hunk 1]
```diff
@@ -0,0 +1,198 @@
+class Chef
+  class Resource
+    class ChocolateyInstaller < Chef::Resource
+      provides :chocolatey_installer
+
+      description "Use the Chocolatey Installer resource to ensure that Choco is installed to your specification. Use the Chocolatey Feature resource to customize your install"
```
Comment (jaymzh) [inline lib/chef/resource/chocolatey_installer.rb:6]: If that's a resource, use it's name, "Use the`chocolatey_feature` resource..."

If it's not, reword this to be more clear.

File: lib/chef/resource/chocolatey_installer.rb [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+            action :upgrade
+            chocolatey_version '2.12.24'
+          end
+          ```
+      DOC
+
+      allowed_actions :install, :uninstall, :upgrade
+
+      property :download_url, String,
+        description: "The URL to download Chocolatey from. This defaults to the value of $env:ChocolateyDownloadUrl, if it is set, and otherwise falls back to the official Chocolatey community repository to download the Chocolatey package. It can be used for offline installation by providing a path to a Chocolatey.nupkg."
```
Comment (jaymzh) [inline lib/chef/resource/chocolatey_installer.rb:48]: We don't usually consult env vars from cookbooks.  It breaks idempotency... Two prior with different environments running this could move the install back and forth between different versions. I highly highly recommend against this.

File: lib/chef/resource/chocolatey_installer.rb [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+      property :chocolatey_version, String,
+        description: "Specifies a target version of Chocolatey to install. By default, the latest stable version is installed. This will use the value in $env:ChocolateyVersion by default, if that environment variable is present. This parameter is ignored if download_url is set."
```
Comment (jaymzh) [inline lib/chef/resource/chocolatey_installer.rb:51]: Same

File: lib/chef/resource/chocolatey_installer.rb [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+      def is_choco_installed?
+        ::File.exist?("#{ENV["ALLUSERSPROFILE"]}\\chocolatey\\bin\\choco.exe")
+      end
+
+      def get_choco_version
+        powershell_exec("choco --version").result
+      end
+
+      def existing_version
```
Comment (jaymzh) [inline lib/chef/resource/chocolatey_installer.rb:82]: Does it conform to gem versions? We have other version helpers if not. Gem versions are very strict.

File: lib/chef/resource/chocolatey_installer.rb [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+        Gem::Version.new(get_choco_version)
+      end
+
+      def define_resource_requirements
+        [ new_resource.proxy_user, new_resource.proxy_password ].each do
```
Comment (jaymzh) [inline lib/chef/resource/chocolatey_installer.rb:87]: I don't think you need this outside loop

File: lib/chef/resource/chocolatey_installer.rb [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+              (!new_resource.proxy_user.nil? && new_resource.proxy_password.nil?) || (new_resource.proxy_user.nil? && !new_resource.proxy_password.nil?)
+            end
+            a.failure_message(Chef::Exceptions::ValidationFailed, "You must specify both a proxy_user and a proxy_password")
+            a.whyrun("Assuming that if you have configured a 'proxy_user' you must also supply a 'proxy_password'")
+          end
+        end
+      end
+
+      action :install, description: "Installs Chocolatey package manager" do
+        unless new_resource.download_url.nil?
```
Comment (jaymzh) [inline lib/chef/resource/chocolatey_installer.rb:99]: You can just do
```suggestion
        if new_resource.download_url
```

Ruby treats that was defined, not nil, not false.

File: lib/chef/resource/chocolatey_installer.rb [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+          powershell_exec("Set-Item -path env:ChocolateyDownloadUrl -Value #{new_resource.download_url}")
+        end
+
+        unless new_resource.chocolatey_version.nil?
```
Comment (jaymzh) [inline lib/chef/resource/chocolatey_installer.rb:103]: Same

File: lib/chef/resource/chocolatey_installer.rb [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+          powershell_exec("Set-Item -path env:ChocolateyVersion -Value #{new_resource.chocolatey_version}")
+        end
+
+        if new_resource.use_native_unzip == true
```
Comment (jaymzh) [inline lib/chef/resource/chocolatey_installer.rb:107]:
```suggestion
        if new_resource.use_native_unzip
```

File: lib/chef/resource/chocolatey_installer.rb [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+          powershell_exec("Set-Item -path env:ChocolateyUseWindowsCompression -Value true")
+        end
+
+        if new_resource.ignore_proxy == true
```
Comment (jaymzh) [inline lib/chef/resource/chocolatey_installer.rb:111]:
```suggestion
        if new_resource.ignore_proxy
```

File: lib/chef/resource/chocolatey_installer.rb [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+          powershell_exec("Set-Item -path env:ChocolateyIgnoreProxy -Value true")
+        end
+
+        unless new_resource.proxy_url.nil?
```
Comment (jaymzh) [inline lib/chef/resource/chocolatey_installer.rb:115]: See above

File: lib/chef/resource/chocolatey_installer.rb [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+          powershell_exec("Set-Item -path env:ChocolateyProxyLocation -Value #{new_resource.proxy_url}")
+        end
+
+        unless !new_resource.proxy_user.nil? && !new_resource.proxy_password.nil?
+          powershell_exec("Set-Item -path env:ChocolateyProxyUser -Value #{new_resource.proxy_user}; Set-Item -path env:ChocolateyProxyPassword -Value #{new_resource.proxy_password}")
+        end
+
+        converge_if_changed do
+          powershell_exec("iex ((New-Object System.Net.WebClient).DownloadString('https://chocolatey.org/install.ps1'))").error!
```
Comment (jaymzh) [inline lib/chef/resource/chocolatey_installer.rb:124]: This downloads it, but it doesn't run the ps1 that's downloaded... So it's not installed... I think.

File: lib/chef/resource/chocolatey_installer.rb [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+        end
+      end
+
+      action :upgrade, description: "Upgrades the Chocolatey package manager" do
+        if !new_resource.chocolatey_version.nil?
```
Comment (jaymzh) [inline lib/chef/resource/chocolatey_installer.rb:129]: Switch this around

If version
  ...
Else
...
End

(Sorry, typing on phone)

File: lib/chef/resource/chocolatey_installer.rb [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+          proposed_version = Gem::Version.new(new_resource.chocolatey_version)
+        else
+          proposed_version = nil
+        end
+
+        unless new_resource.download_url.nil?
```
Comment (jaymzh) [inline lib/chef/resource/chocolatey_installer.rb:135]:
```suggestion
        if new_resource.download_url
```

File: lib/chef/resource/chocolatey_installer.rb [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+          powershell_exec("Set-Item -path env:ChocolateyDownloadUrl -Value #{new_resource.download_url}")
+        end
+
+        unless new_resource.chocolatey_version.nil?
```
Comment (jaymzh) [inline lib/chef/resource/chocolatey_installer.rb:139]: Same

File: lib/chef/resource/chocolatey_installer.rb [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+          powershell_exec("Set-Item -path env:ChocolateyVersion -Value #{new_resource.chocolatey_version}")
+        end
+
+        if new_resource.use_native_unzip == true
```
Comment (jaymzh) [inline lib/chef/resource/chocolatey_installer.rb:143]:
```suggestion
        if new_resource.use_native_unzip
```

File: lib/chef/resource/chocolatey_installer.rb [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+          powershell_exec("Set-Item -path env:ChocolateyUseWindowsCompression -Value true")
+        end
+
+        if new_resource.ignore_proxy == true
```
Comment (jaymzh) [inline lib/chef/resource/chocolatey_installer.rb:147]:
```suggestion
        if new_resource.ignore_proxy
```

File: lib/chef/resource/chocolatey_installer.rb [hunk 17]
```diff
@@ -0,0 +1,198 @@
...
+          powershell_exec("Set-Item -path env:ChocolateyIgnoreProxy -Value true")
+        end
+
+        unless new_resource.proxy_url.nil?
```
Comment (jaymzh) [inline lib/chef/resource/chocolatey_installer.rb:151]: Same

File: lib/chef/resource/chocolatey_installer.rb [hunk 18]
```diff
@@ -0,0 +1,198 @@
...
+          powershell_exec("Set-Item -path env:ChocolateyProxyLocation -Value #{new_resource.proxy_url}")
+        end
+
+        unless !new_resource.proxy_user.nil? && !new_resource.proxy_password.nil?
+          powershell_exec("Set-Item -path env:ChocolateyProxyUser -Value #{new_resource.proxy_user}; Set-Item -path env:ChocolateyProxyPassword -Value #{new_resource.proxy_password}")
+        end
+
+        if !proposed_version.nil? && existing_version < proposed_version
```
Comment (jaymzh) [inline lib/chef/resource/chocolatey_installer.rb:159]:
```suggestion
        if proposed_version && existing_version < proposed_version
```

File: lib/chef/resource/chocolatey_installer.rb [hunk 19]
```diff
@@ -0,0 +1,198 @@
...
+          powershell_exec("Remove-Item -path env:ChocolateyVersion")
+        end
+
+        converge_by("upgrade choco version") do
+          powershell_exec("choco upgrade Chocolatey -y").result
+        end
+      end
+
+      action :uninstall, description: "Uninstall Chocolatey package manager" do
+        converge_by("Uninstall Choco") do
```
Comment (jaymzh) [inline lib/chef/resource/chocolatey_installer.rb:171]: Do the if first and only do the converge_by if there is something to do.

**Review by jaymzh — CHANGES_REQUESTED** [review]

File: lib/chef/resource/chocolatey_installer.rb (hunk 1 above)
Comment (johnmccrae) [reply lib/chef/resource/chocolatey_installer.rb:6]: done

File: lib/chef/resource/chocolatey_installer.rb (hunk 8 above)
Comment (tpowell-progress) [reply lib/chef/resource/chocolatey_installer.rb:107]: I've clearly been dabbling in too many languages that don't have this as a pattern, because I have missed this in 2-3 iterations of this PR.

File: lib/chef/resource/chocolatey_installer.rb (hunk 2 above)
Comment (johnmccrae) [reply lib/chef/resource/chocolatey_installer.rb:48]: I have a question - this defaults to a choco repo if you specify nothing. The point of this url is the use case where a random update needs to be downloaded from an alternate source. It does not need to be idempotent.

File: lib/chef/resource/chocolatey_installer.rb [hunk 20]
```diff
@@ -0,0 +1,198 @@
...
+
+      def existing_version
+        Gem::Version.new(get_choco_version)
+      end
+
+      def define_resource_requirements
+        [ new_resource.proxy_user, new_resource.proxy_password ].each do
+          requirements.assert(:install, :upgrade).each do |a|
+            a.assertion do
+              (!new_resource.proxy_user.nil? && new_resource.proxy_password.nil?) || (new_resource.proxy_user.nil? && !new_resource.proxy_password.nil?)
```
Comment (tpowell-progress) [inline lib/chef/resource/chocolatey_installer.rb:90]: proxy_user.nil?|proxy_password.nil?|evaluate
       ---|       ---|---
       nil|       nil|false
       nil|   not nil|true
   not nil|       nil|true
   not nil|   not nil|false

Comment (tpowell-progress) [inline lib/chef/resource/chocolatey_installer.rb:90]:
```suggestion
              (new_resource.proxy_user.nil? != new_resource.proxy_password.nil?)
```

File: lib/chef/resource/chocolatey_installer.rb (hunk 4 above)
Comment (johnmccrae) [reply lib/chef/resource/chocolatey_installer.rb:82]: BTW, got a pointer for me? Everything I read says to use Gem::Version

File: lib/chef/resource/chocolatey_installer.rb (hunk 20 above)
Comment (johnmccrae) [reply lib/chef/resource/chocolatey_installer.rb:90]: OOh, thanks

File: lib/chef/resource/chocolatey_installer.rb (hunk 6 above)
Comment (johnmccrae) [reply lib/chef/resource/chocolatey_installer.rb:99]: done

File: lib/chef/resource/chocolatey_installer.rb (hunk 8 above)
Comment (johnmccrae) [reply lib/chef/resource/chocolatey_installer.rb:107]: done

File: lib/chef/resource/chocolatey_installer.rb (hunk 10 above)
Comment (johnmccrae) [reply lib/chef/resource/chocolatey_installer.rb:115]: done

File: lib/chef/resource/chocolatey_installer.rb (hunk 11 above)
Comment (johnmccrae) [reply lib/chef/resource/chocolatey_installer.rb:124]: ? I ran this on bare metal and it worked?

Comment (johnmccrae) [reply lib/chef/resource/chocolatey_installer.rb:124]: testing again

File: lib/chef/resource/chocolatey_installer.rb (hunk 4 above)
Comment (tpowell-progress) [reply lib/chef/resource/chocolatey_installer.rb:82]: If we're just comparing two *specific* gem versions, `Gem::Version` sounds good to me. If we're looking to have more general version specifiers (`~>`, etc...), it might still work, but we might want to consider how the user would expect things to work in context.

File: lib/chef/resource/chocolatey_installer.rb (hunk 11 above)
Comment (johnmccrae) [reply lib/chef/resource/chocolatey_installer.rb:124]: I see what you're asking now. the IEX at the beginning of the block is a short-cut for Invoke-Expression. So the IEX calls the block which downloads the installer ps1 file then executes it.

File: lib/chef/resource/chocolatey_installer.rb (hunk 4 above)
Comment (johnmccrae) [reply lib/chef/resource/chocolatey_installer.rb:82]: I'm a total dork, ignore me. I understand now. I was overthinking this.

File: lib/chef/resource/chocolatey_installer.rb (hunk 2 above)
Comment (tpowell-progress) [reply lib/chef/resource/chocolatey_installer.rb:48]: So, without involving the environment variable, we would have the default location and this override? Seems like having the environment variable as yet another backup configuration location might be overkill, anyway. Just default to the standard url and have this as an override.

File: lib/chef/resource/chocolatey_installer.rb (hunk 11 above)
Comment (tpowell-progress) [inline lib/chef/resource/chocolatey_installer.rb:124]:
```suggestion
          # note that Invoke-Expression is being called on the downloaded script (outer parens), not triggering the script download (inner parens)
          powershell_exec("Invoke-Expression ((New-Object System.Net.WebClient).DownloadString('https://chocolatey.org/install.ps1'))").error!
```

File: lib/chef/resource/chocolatey_installer.rb (hunk 2 above)
Comment (tpowell-progress) [reply lib/chef/resource/chocolatey_installer.rb:48]: Per our discussion, this is actually being used directly by the Chocolatey `install.ps1` script and the environment variable can be used to redirect the standard installer to an alternate location.

File: lib/chef/resource/chocolatey_installer.rb (hunk 3 above)
Comment (tpowell-progress) [reply lib/chef/resource/chocolatey_installer.rb:51]: Per our discussion, this is actually being used directly by the Chocolatey `install.ps1` script and the environment variable can be used to redirect the standard installer to an alternate _version_.

**Review by tpowell-progress — CHANGES_REQUESTED** [review]
Logic simplification

File: lib/chef/resource/chocolatey_installer.rb [hunk 21]
```diff
@@ -0,0 +1,195 @@
...
+      end
+
+      def existing_version
+        Gem::Version.new(get_choco_version)
+      end
+
+      def define_resource_requirements
+        requirements.assert(:install, :upgrade).each do |a|
+          a.assertion do
+            (new_resource.proxy_user.nil? != new_resource.proxy_password.nil?)
```
Comment (tpowell-progress) [inline lib/chef/resource/chocolatey_installer.rb:89]: Go ahead and remove the parens and comment that we want an error if only one of (proxy_user, proxy_password) are true.

File: lib/chef/resource/chocolatey_installer.rb (hunk 11 above)
Comment (johnmccrae) [reply lib/chef/resource/chocolatey_installer.rb]: done

**Review by tpowell-progress — APPROVED** [review]
Re-approving after talking through the env var concerns on the Community PR review call.

File: lib/chef/resource/chocolatey_installer.rb (hunk 2 above)
Comment (tpowell-progress) [reply lib/chef/resource/chocolatey_installer.rb:48]: @johnmccrae can we made the description more clear that it's the choco installer itself looking at these variables?

Comment (johnmccrae) [reply lib/chef/resource/chocolatey_installer.rb]: done and spelling error corrected too

File: lib/chef/resource/chocolatey_installer.rb (hunk 21 above)
Comment (johnmccrae) [reply lib/chef/resource/chocolatey_installer.rb]: done
